src/map_to_curve.rs

```diff
@@ -56,6 +56,35 @@ impl GroupWithMapToCurve for RistrettoPoint {
     }
 }
 
+#[cfg(feature = "p256")]
+impl GroupWithMapToCurve for p256_::ProjectivePoint {
+    const SUITE_ID: usize = 0x0003;
+
+    fn map_to_curve<H: Hash>(msg: &[u8], dst: &[u8]) -> Result<Self, ProtocolError> {
+        // https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-11#section-8.2
+        // `P256_XMD:SHA-256_SSWU_RO_` has an `L` of `48`
+        const L: usize = 48;
+
+        // https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-11#section-3
+        // `hash_to_curve` calls `hash_to_field` with a `count` of `2`
+        // https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-11#section-5.3
+        // `hash_to_field` calls `expand_message` with a `len_in_bytes` of `count * L`
+        let uniform_bytes = expand_message_xmd::<H>(msg, dst, 2 * L)?;
+
+        <Self as Group>::hash_to_curve(&GenericArray::clone_from_slice(&uniform_bytes[..]))
+            .map_err(ProtocolError::from)
+    }
```

Comment:
This here seems to be the crux of the problem for me. Following the draft, I am trying to implement [`hash_to_curve`](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-11#section-3), which looks like this:
```
   Steps:
   1. u = hash_to_field(msg, 2)
   2. Q0 = map_to_curve(u[0])
   3. Q1 = map_to_curve(u[1])
   4. R = Q0 + Q1              # Point addition
   5. P = clear_cofactor(R)
   6. return P
```
But the current implementation was for ristretto255 only, which implemented [`hash_to_ristretto255`](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-11#appendix-B):
```
   Steps:
   1. uniform_bytes = expand_message(msg, DST, 64)
   2. P = ristretto255_map(uniform_bytes)
   3. return P
```
The way we named and ordered things doesn't seem to entirely align with the spec:
- `GroupWithMapToCurve` has a method `map_to_curve`
- `Group` has a method `hash_to_curve`
- `GroupWithMapToCurve::map_to_curve` calls `Group::hash_to_curve`
- Ristretto255:
  - Step 1 (`uniform_bytes = expand_message(msg, DST, 64)`): in `GroupWithMapToCurve::map_to_curve`
  - Step 2 & 3 (`ristretto255_map(uniform_bytes)`): in `Group::hash_to_curve`
- P-256:
  - Step 1 (`hash_to_field(msg, 2)`): split in `GroupWithMapToCurve::map_to_curve` and `Group::hash_to_curve`
  - Step 2 - 6 (`Q0 = map_to_curve(u[0])`): in `Group::hash_to_curve`

The reason I had to split up the P-256 implementation like this, is because I didn't want to continuously re-en/decode types between calls.

What is the reason it is like that? What is the point of `GroupWithMapToCurve`?

Reply:
Indeed, the naming and separation of Group vs. GroupWithMapToCurve no longer makes sense. (It was like this back when the spec was a much older version, but now it's just confusing.) I would be in favor of simply consolidating the two traits, since there no longer needs to be a separation.

Reply:
I will do this in a follow up PR.